keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,285 @@
+1. The COSI controller deletes the Bucket's bound BucketContent object.
+1. The COSI controller removes the finalizer from the Bucket, allowing it and the child Secret to be garbage collected.
+
+##  Custom Resource Definitions
+
+#### Bucket
+
+A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the Bucket is "bound" to the corresponding BucketContent. The is used to prevent further binds to the BucketContent.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  bucketName: [3]
+  generateBucketName: [4]
+  bucketClassName: [5]
+  secretName: [6]
+  protocol: [7]
+	accessMode: {"ro", "rw"} [8]
+status:
+  bucketContentName: [9]
+  phase:
+  conditions: 
+```
+1. `labels`: COSI controller adds the label to its managed resources to easy CLI GET ops.  Value is the driver name returned by GetDriverInfo() rpc*.
+1. `finalizers`: COSI controller adds the finalizer to defer Bucket deletion until backend deletion ops succeed.
+1. `bucketName`: Desired name of the bucket to be created**.
+1. `generateBucketName`: Desired prefix to a randomly generated bucket name. Mutually exclusive with `bucketName`**.
+1. `bucketClassName`: Name of the target BucketClass.
+1. `secretName`: Desired name for user's credential Secret. Fails on name collisions. Deterministic names allow for a single manifest workflow.
+1. `protocol`: String array of protocols (e.g. s3, gcs, swift, etc.) requested by the user.  Used in matching Buckets to BucketClasses and ensuring compatibility with backing object stores.
+1. `accessMode`:  The requested level of access provided to the returned access credentials.
+1. `bucketContentName`: Name of a bound BucketContent
+
+> \* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+
+> ** Ignored in Brownfield and Static operations
+
+#### BucketContent
+
+A cluster-scoped resource representing an object store bucket. The BucketContent is expected to store stateful data relevant to bucket deprovisioning. The BucketContent is bound to the Bucket in a 1:1 mapping.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketContent
+Metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  bucketClassName: [4]
+  supportedProtocols: [5]
+  releasePolicy: {"Delete", "Retain"} [6]
+  bucketRef: [7]
+    name:
+    namespace:
+  secretRef: [8]
+  accessMode: {"rw", "ro"} [9]
+status:
+  bucketAttributes: <map[string]string> [10]
+  phase: {"Bound", "Released", "Failed", "Errored"} [11]
+  conditions:
+```
+1. `name`: Generated in the pattern of `“bucket-”<BUCKET-NAMESPACE>"-"<BUCKET-NAME>`
```

Comment:
what are the k8s truncation rules? What do we do if the `bucket`'s name + ns exceeds 63 chars?

Reply:
Going from memory, the base prefix is truncated by the length of the randomized suffix.